include/sycl_ext_complex.hpp

```diff
@@ -358,628 +354,343 @@ template <class _Tp> class complex {
+    value_type __c = __w.real();
+    value_type __d = __w.imag();
+    value_type __ac = __a * __c;
+    value_type __bd = __b * __d;
+    value_type __ad = __a * __d;
+    value_type __bc = __b * __c;
+    value_type __x = __ac - __bd;
+    value_type __y = __ad + __bc;
+    if (sycl::isnan(__x) && sycl::isnan(__y)) {
+      bool __recalc = false;
+      if (sycl::isinf(__a) || sycl::isinf(__b)) {
+        __a = sycl::copysign(sycl::isinf(__a) ? value_type(1) : value_type(0), __a);
+        __b = sycl::copysign(sycl::isinf(__b) ? value_type(1) : value_type(0), __b);
+        if (sycl::isnan(__c))
+          __c = sycl::copysign(value_type(0), __c);
+        if (sycl::isnan(__d))
+          __d = sycl::copysign(value_type(0), __d);
+        __recalc = true;
+      }
+      if (sycl::isinf(__c) || sycl::isinf(__d)) {
+        __c = sycl::copysign(sycl::isinf(__c) ? value_type(1) : value_type(0), __c);
+        __d = sycl::copysign(sycl::isinf(__d) ? value_type(1) : value_type(0), __d);
+        if (sycl::isnan(__a))
+          __a = sycl::copysign(value_type(0), __a);
+        if (sycl::isnan(__b))
+          __b = sycl::copysign(value_type(0), __b);
+        __recalc = true;
+      }
+      if (!__recalc && (sycl::isinf(__ac) || sycl::isinf(__bd) ||
+                        sycl::isinf(__ad) || sycl::isinf(__bc))) {
+        if (sycl::isnan(__a))
+          __a = sycl::copysign(value_type(0), __a);
+        if (sycl::isnan(__b))
+          __b = sycl::copysign(value_type(0), __b);
+        if (sycl::isnan(__c))
+          __c = sycl::copysign(value_type(0), __c);
+        if (sycl::isnan(__d))
+          __d = sycl::copysign(value_type(0), __d);
+        __recalc = true;
+      }
+      if (__recalc) {
+        __x = value_type(INFINITY) * (__a * __c - __b * __d);
+        __y = value_type(INFINITY) * (__a * __d + __b * __c);
+      }
     }
-    if (!__recalc && (sycl::isinf(__ac) || sycl::isinf(__bd) ||
-                      sycl::isinf(__ad) || sycl::isinf(__bc))) {
-      if (sycl::isnan(__a))
-        __a = sycl::copysign(_Tp(0), __a);
-      if (sycl::isnan(__b))
-        __b = sycl::copysign(_Tp(0), __b);
-      if (sycl::isnan(__c))
-        __c = sycl::copysign(_Tp(0), __c);
-      if (sycl::isnan(__d))
-        __d = sycl::copysign(_Tp(0), __d);
-      __recalc = true;
+    return complex<value_type>(__x, __y);
+  }
+  _SYCL_EXT_CPLX_INLINE_VISIBILITY friend complex<value_type>
+  operator*(const complex<value_type> &__x, const value_type &__y) {
+    complex<value_type> __t(__x);
+    __t *= __y;
+    return __t;
+  }
+  _SYCL_EXT_CPLX_INLINE_VISIBILITY friend complex<value_type>
+  operator*(const value_type &__x, const complex<value_type> &__y) {
+    complex<value_type> __t(__y);
+    __t *= __x;
+    return __t;
+  }
+
+  _SYCL_EXT_CPLX_INLINE_VISIBILITY friend complex<value_type>
+  operator/(const complex<value_type> &__z, const complex<value_type> &__w) {
+    int __ilogbw = 0;
+    value_type __a = __z.real();
```

Comment:
Can't this use the `.__x` directly, since it's a friend? I am not sure there is any performance benefit, I expect most compilers well inline the `.real`, but it was mentioned in draft of the paper.

Reply:
You are right, nice catch